storage/sqlite/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+            logging::log::error!("Error: transaction rollback failed: {}", err);
+        }
+    }
+}
+
+impl<'s, 'i> backend::PrefixIter<'i> for DbTx<'s> {
+    type Iterator = PrefixIter;
+
+    fn prefix_iter<'t: 'i>(
+        &'t self,
+        idx: DbIndex,
+        prefix: Data,
+    ) -> storage_core::Result<Self::Iterator> {
+        // TODO check if prefix.is_empty()
+        // TODO Perform the filtering in the SQL query itself
+        let mut stmt = self
+            .connection
+            .prepare_cached(self.queries[idx].prefix_iter_query.as_str())
+            .map_err(process_sqlite_error)?;
+
+        let mut rows = stmt.query(()).map_err(process_sqlite_error)?;
+
+        // TODO Move the statement/rows in to the PrefixIter
+        let mut kv = Vec::new();
+        while let Some(row) = rows.next().map_err(process_sqlite_error)? {
+            let key = row.get::<usize, Vec<u8>>(0).map_err(process_sqlite_error)?;
+            if key.starts_with(&prefix) {
+                let value = row.get::<usize, Vec<u8>>(1).map_err(process_sqlite_error)?;
+                kv.push((key, value));
+            }
+        }
+        let kv_iter = kv.into_iter();
+
+        Ok(PrefixIter::new(kv_iter, prefix))
+    }
+}
+
+impl backend::ReadOps for DbTx<'_> {
+    fn get(&self, idx: DbIndex, key: &[u8]) -> storage_core::Result<Option<Cow<[u8]>>> {
+        let mut stmt = self
+            .connection
+            .prepare_cached(self.queries[idx].get_query.as_str())
+            .map_err(process_sqlite_error)?;
+
+        let params = (key,);
+        let res = stmt
+            .query_row(params, |row| row.get::<usize, Vec<u8>>(0))
+            .optional()
+            .map_err(process_sqlite_error)?;
+        let res = res.map(|v| v.into());
+        Ok(res)
+    }
+}
+
+impl backend::WriteOps for DbTx<'_> {
+    fn put(&mut self, idx: DbIndex, key: Data, val: Data) -> storage_core::Result<()> {
+        let mut stmt = self
+            .connection
+            .prepare_cached(self.queries[idx].put_query.as_str())
+            .map_err(process_sqlite_error)?;
+
+        let params = (key, val);
+        let _res = stmt.execute(params).map_err(process_sqlite_error)?;
+
+        Ok(())
+    }
+
+    fn del(&mut self, idx: DbIndex, key: &[u8]) -> storage_core::Result<()> {
+        let mut stmt = self
+            .connection
+            .prepare_cached(self.queries[idx].delete_query.as_str())
+            .map_err(process_sqlite_error)?;
+
+        let params = (key,);
+        let _res = stmt.execute(params).map_err(process_sqlite_error)?;
+
+        Ok(())
+    }
+}
+
+impl backend::TxRo for DbTx<'_> {}
+
+impl backend::TxRw for DbTx<'_> {
+    fn commit(self) -> storage_core::Result<()> {
+        self.commit_transaction()
+    }
+}
+
+#[derive(Clone)]
+pub struct SqliteImpl {
+    /// Handle to an Sqlite database connection
+    connection: Arc<Mutex<Connection>>,
+
+    /// List of sql queries
+    queries: SqliteQueries,
```

Comment:
Minor nitpick: Both `queries` and `connection` can just share one `Arc`, since they are always cloned together. One reference counter is enough. It does however mean introducing an intermediate struct to keep it readable so I don't mind either way.

Reply:
We could do that. Something like [this](https://github.com/mintlayer/mintlayer-core/commit/dce92f9bb71f79be6161a4d34302efd707df878b)?